src/shogun/machine/TypedMachine.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Shubham Shukla
+ */
+
+#ifndef _TYPEDMACHINE_H__
+#define _TYPEDMACHINE_H__
+
+#include <shogun/lib/config.h>
+
+#include <shogun/base/progress.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/common.h>
+#include <shogun/machine/LinearMachine.h>
+
+namespace shogun
+{
+	class CFeatures;
+	class CLabels;
+	
+	template <class P, class T>
+	class CTypedMachine : public T
+	{
+	public:
+		/** Default constructor */
+		CTypedMachine() : T()
+		{
+		}
+
+		virtual ~CTypedMachine()
+		{
+		}
+
+	protected:
+		virtual bool train_machine(CFeatures* data = NULL)
+		{
+				if (!data)
+				{	
+					REQUIRE(this->features, "No features provided.\n")	
+					REQUIRE(this->features->get_feature_class() == C_DENSE,	
```

Comment:
So this class is supposed to JUST work with dense features? Then we need one for string and sparse, etc as well....

Reply:
we can add more cases like we did in macro version and then throw error if nothing matches

Reply:
i think the problem here is that we force the subclasses to implement the templated train method. So if we add string here, we will get a compile error when it is not implemented for strings....or?